compiler/rustc_mir/src/interpret/intern.rs

```diff
@@ -2,12 +2,23 @@
 //!
 //! After a const evaluation has computed a value, before we destroy the const evaluator's session
 //! memory, we need to extract all memory allocations to the global memory pool so they stay around.
+//!
+//! In principle, this is not very complicated: we recursively walk the final value, follow all the
+//! pointers, and move all reachable allocations to the global `tcx` memory. The only complication
+//! is picking the right mutability for the allocations in a `static` initializer: we want to make
+//! as many allocations as possible immutable so LLVM can put them into read-only memory. At the
+//! same time, we need to make memory that could be mutated by the program mutable to avoid
+//! incorrect compilations. To achieve this, we do a type-based traversal of the final value,
+//! tracking mutable and shared references and `UnsafeCell` to determine the current mutability.
```

Comment:
Does that mean we could theoretically mark all static allocations as mutable? Is this pass slow to run, would that have any significant speedup?

Reply:
It's the other way around, we can just mark all allocations in constants as immutable. I'm not sure whether we'd get a speedup though.

Reply:
> Does that mean we could theoretically mark all static allocations as mutable? Is this pass slow to run, would that have any significant speedup?

For allocations in a `static`, yes, And then many people will complain that their read-only `static` are not put into read-only memory. So that's not really an option.